core/src/test/scripts/unit_tests_validate_data.py

```diff
@@ -941,25 +965,184 @@ def test_missing_row_gsva(self):
-                         record.getMessage())
+        self.assertEqual(record.line_number, 1)
+        self.assertIn('headers', record.getMessage().lower())
+        self.assertIn('different', record.getMessage().lower())
+        record = next(record_iterator)
+        self.assertIn('invalid', record.getMessage().lower())
+        self.assertIn('column', record.getMessage().lower())
         return
+                                    
+    def test_missing_row_effectiveconcentration(self):
+        self.logger.setLevel(logging.ERROR)
+
+        # reset state of the TreatmentWiseFileValidator 
+        # class to prevent carry-over from other tests
+        _resetMultipleFileHandlerClassVars()
+
+        ### Error should appear when the second file is validated
+        record_list1 = self.validate('study_es_0/data_treatment_ic50.txt',
+                                    TreatmentMultiFileTestCase.DummyTreatmentValidator)
+        record_list2 = self.validate('data_treatment_ic50_missing_row.txt',
+                                    TreatmentMultiFileTestCase.DummyTreatmentValidator)
 
-    # TODO: test other subclasses of FeatureWiseValidator
+        self.assertEqual(len(record_list1), 0)
+        self.assertEqual(len(record_list2), 1)
+        for record in record_list2:
+            self.assertEqual(record.levelno, logging.ERROR)
+        self.assertEqual('Feature columns in data files are not equal.', record.getMessage())
+        return
+
+# ------------------ end treatment multifile consistency test ------------------
+
+# -------------- treatment effective concentration validator test --------------
+
+class EffectiveConcentrationValidatorTestCase(unittest.TestCase):
+
+    def test_empty_cell_issues_error(self):
+        passConcValue("").error.assert_called()
+
+    def test_NA_value_is_allowed(self):
+        passConcValue("NA").error.assert_not_called()
+
+    def test_NAN_value_issues_error(self):
+        passConcValue("NAN").error.assert_called()
+        passConcValue("nan").error.assert_called()
+        passConcValue("NaN").error.assert_called()
+
+    def test_Inf_value_issues_error(self):
+        passConcValue("Inf").error.assert_called()
+        passConcValue("inf").error.assert_called()
+        passConcValue("Infinite").error.assert_called()
+        passConcValue("infinite").error.assert_called()
+
+    def test_no_decimals_issues_warning(self):
+        passConcValue("10").warning.assert_called()
+        passConcValue("00010").warning.assert_called()
+    
+    def test_zero_issues_error(self):
+        passConcValue("0").error.assert_called()
+        passConcValue("0.000").error.assert_called()
+        passConcValue("00000").error.assert_called()
+
+    def test_leadingzeros_are_allowed(self):
+        passConcValue("0000.123").error.assert_not_called()
+
+    def test_negativenumber_issues_error(self):
+        passConcValue("-234234.234").error.assert_called()
```

Comment:
I wouldn't block negative numbers now that we know that log values could be one of the use cases.

Reply:
The idea was that concentrations can never be negative. I plan to implement two treatment data types. One is TREATMENT_CONCENTRATION (must be > 0), the other TREATMENT_CONTINUOUS (any numerical value). When a neg.log of concentration is passed the TREATMENT_CONTINUOUS data type should be used. Alternatively, we only create a TREATMENT_CONTINUOUS that is applied to every data file. You prefer the latter?